File: apps/apple/TeslaSync/feature-views/KioskOverlay.Model.swift
//
//  KioskOverlay.Model.swift
//  TeslaSync — P4 feature view · 0124 · KioskOverlay (Apple)
//
//  The pure, host-free projection of a `KioskOverlay`'s inputs into the structural
//  decisions the view renders — the native parity of the web component's render
//  branches (web/src/features/dashboard/components/KioskOverlay.tsx):
//
//    • the dim layer + its `opacity: 1 - config.dimLevel` (CSS-clamped to 0…1),
//    • the cursor-hide decision (`isCursorHidden`),
//    • the clock decision + its corner (`config.showClock` / `config.clockPosition`),
//    • the rotation-indicator gate (`dashboardCount > 1 && config.rotateInterval > 0`)
//      and the active-dot test (`i === currentIndex`),
//    • the locale-aware clock formatters (ports of `useDateFormat`'s `formatTime`
//      and `formatDateWithDay`),
//    • the 3-second exit auto-hide delay (`setTimeout(…, 3000)`).
//
//  `KioskOverlay` owns no data — exactly like the web component, which receives
//  everything as props and fetches nothing — so the loading / empty / error /
//  stale / offline states belong to whatever surface embeds the overlay (the Kiosk
//  page), not to the overlay itself. The branches the web source actually carries
//  are reproduced here in full and are unit-tested without a rendering host.
//
//  Keeping these decisions in `Equatable` value types lets the XCTest suite cover
//  every configuration (and the accessibility / i18n policy) without a snapshot
//  host — the same approach the sibling presentational surfaces use.
//

import Foundation
import SwiftUI

// MARK: - Surface identity

/// Stable, non-identifying identity for the `KioskOverlay` feature view. The slug
/// is the value emitted with the P1/S11 `view.opened` diagnostics contract and is
/// referenced by both the view and its tests so the two never drift.
public enum KioskOverlaySurface {
    /// The diagnostics slug emitted with the `view.opened` event.
    public static let slug = "KioskOverlay"

    /// Reports the surface becoming visible. This is the exact code path the view
    /// runs from its `.task`, factored out so it is unit-testable without a
    /// rendering host.
    public static func reportOpen(to telemetry: any KioskOverlayTelemetry) {
        telemetry.viewOpened(surface: slug)
    }
}

// MARK: - Localization facade (P1/S10) — web `t(key, default)`

/// Resolves the surface's strings by key with the web English fallback, so the
/// view holds no hardcoded literals. Keys live in the "KioskOverlay" table, folded
/// into the app `Localizable.xcstrings` catalog at integration time; the
/// per-surface table keeps each parallel surface prompt self-contained.
public enum KioskOverlayStrings {
    /// The per-surface strings table name (file `KioskOverlay.strings`).
    public static let table = "KioskOverlay"

    /// The localized value for `key`, falling back to the web English literal.
    public static func string(_ key: String, _ fallback: String) -> String {
        NSLocalizedString(key, tableName: table, bundle: .main, value: fallback, comment: "")
    }

    /// `string(_:_:)` wrapped as a `Text` (rendered verbatim so the resolved value
    /// is shown exactly).
    public static func text(_ key: String, _ fallback: String) -> Text {
        Text(verbatim: string(key, fallback))
    }
}

// MARK: - Clock corner (web `config.clockPosition`)

/// The corner the kiosk clock pins to, mirroring the web
/// `'top-left' | 'top-right' | 'bottom-left' | 'bottom-right'` union. The raw value
/// is the exact web string so a backend / config payload round-trips losslessly.
public enum KioskClockPosition: String, CaseIterable, Sendable {
    case topLeft = "top-left"
    case topRight = "top-right"
    case bottomLeft = "bottom-left"
    case bottomRight = "bottom-right"

    /// Decodes from the web string, defaulting to `.bottomRight` for an unknown
    /// value — parity with `DEFAULT_KIOSK_CONFIG.clockPosition`.
    public init(webValue: String) {
        self = KioskClockPosition(rawValue: webValue) ?? .bottomRight
    }

    /// The frame alignment that pins the clock box to its corner (web `top-4 left-4`
    /// … `bottom-4 right-4`).
    public var alignment: Alignment {
        switch self {
        case .topLeft: .topLeading
        case .topRight: .topTrailing
        case .bottomLeft: .bottomLeading
        case .bottomRight: .bottomTrailing
        }
    }

    /// The text-stack alignment, so the time/date hug the same screen edge as the
    /// corner (a native polish over the web's always-left-aligned text).
    public var horizontalAlignment: HorizontalAlignment {
        switch self {
        case .topLeft, .bottomLeft: .leading
        case .topRight, .bottomRight: .trailing
        }
    }
}

// MARK: - Overlay configuration (web `config: KioskConfig`)

/// The subset of the web `KioskConfig` the overlay actually reads (`dimLevel`,
/// `showClock`, `clockPosition`, `rotateInterval`). The remaining `KioskConfig`
/// fields drive the kiosk *engine* (cursor/dim timers, rotation, opacity), not this
/// presentational overlay, so they are out of this surface's contract. Defaults
/// mirror `DEFAULT_KIOSK_CONFIG`.
public struct KioskOverlayConfig: Equatable, Sendable {
    /// Web `dimLevel` (0…1). The dim layer renders at `1 - dimLevel`.
    public var dimLevel: Double
    /// Web `showClock`.
    public var showClock: Bool
    /// Web `clockPosition`.
    public var clockPosition: KioskClockPosition
    /// Web `rotateInterval` (seconds). The rotation indicator only shows when > 0.
    public var rotateInterval: Double

    public init(
        dimLevel: Double = 0.5,
        showClock: Bool = true,
        clockPosition: KioskClockPosition = .bottomRight,
        rotateInterval: Double = 30
    ) {
        self.dimLevel = dimLevel
        self.showClock = showClock
        self.clockPosition = clockPosition
        self.rotateInterval = rotateInterval
    }
}

// MARK: - Clock formatting (port of `useDateFormat`)

/// Locale + timezone-aware clock formatters, ports of the web `useDateFormat`
/// helpers used by the overlay:
///
///   • `formatTime`        → `toLocaleTimeString({ hour: '2-digit', minute: '2-digit' })`
///     (24h or 12h per locale) — template `"jjmm"`.
///   • `formatDateWithDay` → `toLocaleDateString({ weekday: 'short', month: 'short',
///     day: 'numeric' })` ("Fri, Apr 4") — template `"EEEMMMd"`.
///
/// `setLocalizedDateFormatFromTemplate` reorders the fields for the active locale,
/// matching `Intl`'s behaviour, so a 24-hour locale yields "14:30" and a 12-hour
/// locale "02:30 PM" from the same call.
public enum KioskClock {
    /// The current wall-clock time, formatted like the web `formatTime` (2-digit
    /// hour + minute, locale 12/24h).
    public static func formatTime(
        _ date: Date,
        locale: Locale = .current,
        timeZone: TimeZone = .current
    ) -> String {
        let formatter = DateFormatter()
        formatter.locale = locale
        formatter.timeZone = timeZone
        formatter.setLocalizedDateFormatFromTemplate("jjmm")
        return formatter.string(from: date)
    }

    /// The current date, formatted like the web `formatDateWithDay` (short weekday +
    /// short month + numeric day).
    public static func formatDateWithDay(
        _ date: Date,
        locale: Locale = .current,
        timeZone: TimeZone = .current
    ) -> String {
        let formatter = DateFormatter()
        formatter.locale = locale
        formatter.timeZone = timeZone
        formatter.setLocalizedDateFormatFromTemplate("EEEMMMd")
        return formatter.string(from: date)
    }
}

// MARK: - Exit affordance timing (web `setTimeout(…, 3000)`)

/// Timing constants for the exit affordance reveal/auto-hide, isolated so the
/// 3-second parity with the web `setShowExit(false)` timer is asserted by a test
/// rather than buried in the view.
public enum KioskOverlayExit {
    /// How long the exit control stays revealed after an interaction before it fades
    /// back out (web `setTimeout(() => setShowExit(false), 3000)`).
    public static let autoHideDelay: Duration = .seconds(3)
}

// MARK: - Presentation (pure projection of the inputs → render config)

/// The pure, `Equatable` projection of a `KioskOverlay`'s inputs into the render
/// decisions: whether/how strongly to dim, whether to hide the pointer, whether and
/// where to show the clock, and whether/how to render the rotation indicator.
public struct KioskOverlayPresentation: Equatable, Sendable {
    /// Web `isDimmed` — gates the dim layer entirely.
    public let isDimmed: Bool
    /// The dim layer opacity, web `1 - config.dimLevel`, CSS-clamped to `0…1`
    /// (non-finite input ⇒ `0`, i.e. no dimming).
    public let dimOpacity: Double
    /// Web `isCursorHidden` — the view hides the macOS pointer when `true` (a no-op
    /// on touch platforms, which have no system pointer).
    public let hidesPointer: Bool
    /// Web `config.showClock`.
    public let showsClock: Bool
    /// Web `config.clockPosition`.
    public let clockPosition: KioskClockPosition
    /// Web `dashboardCount > 1 && config.rotateInterval > 0`.
    public let showsRotationIndicator: Bool
    /// The number of rotation dots (web `dashboardCount`), floored at `0`.
    public let dashboardCount: Int
    /// The active rotation index (web `currentIndex`).
    public let currentIndex: Int

    public init(
        config: KioskOverlayConfig,
        isDimmed: Bool,
        isCursorHidden: Bool,
        dashboardCount: Int,
        currentIndex: Int
    ) {
        self.isDimmed = isDimmed
        dimOpacity = Self.dimOpacity(dimLevel: config.dimLevel)
        hidesPointer = isCursorHidden
        showsClock = config.showClock
        clockPosition = config.clockPosition
        showsRotationIndicator = dashboardCount > 1 && config.rotateInterval > 0
        self.dashboardCount = Swift.max(0, dashboardCount)
        self.currentIndex = currentIndex
    }

    /// The dim opacity, web `1 - dimLevel`, clamped to `0…1`. A non-finite
    /// `dimLevel` resolves to `0` so bad config never blacks out the screen.
    public static func dimOpacity(dimLevel: Double) -> Double {
        guard dimLevel.isFinite else { return 0 }
        return Swift.min(Swift.max(1 - dimLevel, 0), 1)
    }

    /// The dot indices to render (web `Array.from({ length: dashboardCount })`).
    public var dotIndices: Range<Int> {
        0 ..< dashboardCount
    }

    /// Whether the dot at `index` is the active one (web `i === currentIndex`).
    public func isActiveDot(_ index: Int) -> Bool {
        index == currentIndex
    }

    /// The diagnostics slug this presentation belongs to.
    public var surfaceSlug: String {
        KioskOverlaySurface.slug
    }
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/KioskOverlay/KioskOverlayModel.kt
// Pure, framework-free model + projection for the KioskOverlay feature view — the native analogue of
// everything the web component derives before returning JSX
// (web/src/features/dashboard/components/KioskOverlay.tsx). No Compose, no Android, no HTTP: every
// declaration here is unit-tested off-device in the :android:testReleaseUnitTest gate, keeping the
// composable a thin render layer.
//
// KioskOverlay is a purely presentational surface — the web component takes its `config` (a subset of the
// localStorage-backed KioskConfig), `isDimmed`, `isCursorHidden`, `dashboardCount`, `currentIndex`, and an
// `onExit` callback as props from the owning Dashboard page (which owns `useKioskMode` + `useDateFormat`),
// so this surface binds NO data hook and issues NO network call. As in the sibling BatteryPill / StatusHeader
// ports, the cache-then-network lifecycle (loading / empty / error / stale / offline) lives on the owning
// page, not here; modelling those phases would invent behaviour the web spec does not have (drift). The
// branches the web source actually defines are the complete state set this surface renders, and each is
// projected here:
//   1. the dim layer (web `isDimmed && <div … style={{ opacity: 1 - config.dimLevel }}>`),
//   2. the cursor-hidden layer (web `isCursorHidden`; an invisible, aria-hidden CSS injection — it has no
//      visible surface on the web and none on Android, where there is no persistent pointer to hide),
//   3. the clock (web `config.showClock`) in one of four corners, ticking once per second,
//   4. the rotation dots (web `dashboardCount > 1 && config.rotateInterval > 0`) with the active dot wider,
//   5. the exit affordance (always present, revealed by interaction for ~3s).
//
// [KioskOverlayConfig] mirrors the persisted KioskConfig wire shape (camelCase keys, all defaulted to the
// web `DEFAULT_KIOSK_CONFIG`) so the projection can run straight off the cached/persisted JSON; only the
// five fields this surface consumes are modelled, and a decoder must ignore the remaining KioskConfig keys.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/KioskOverlay — the P3 prompt's allowed-files path) cannot form a valid Kotlin
// package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling BatteryPill / StatusHeader surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.kioskoverlay

import io.teslasync.shared.core.diagnostics.Logger
import kotlinx.serialization.Serializable
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import java.util.Locale

/**
 * The four clock anchors the web component supports — the native analogue of the
 * `config.clockPosition === 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right'` ternary that picks
 * the Tailwind corner utility (`top-4 left-4`, …). The composable maps each onto a Compose `Alignment`.
 */
enum class KioskClockPosition {
    TopLeft,
    TopRight,
    BottomLeft,
    BottomRight,
    ;

    companion object {
        /** The web default (`DEFAULT_KIOSK_CONFIG.clockPosition`). */
        val DEFAULT: KioskClockPosition = BottomRight

        /**
         * Parse the persisted wire value (web `'top-left'` / `'top-right'` / `'bottom-left'` /
         * `'bottom-right'`). An unknown or absent value falls back to [DEFAULT], mirroring the web spread
         * over `DEFAULT_KIOSK_CONFIG` that backfills any missing field.
         */
        fun fromWire(value: String?): KioskClockPosition =
            when (value?.trim()?.lowercase(Locale.ROOT)) {
                "top-left" -> TopLeft
                "top-right" -> TopRight
                "bottom-left" -> BottomLeft
                "bottom-right" -> BottomRight
                else -> DEFAULT
            }
    }
}

/**
 * The subset of the web `KioskConfig` this overlay reads, mirroring the persisted localStorage wire shape
 * (`useKioskMode.ts`). All fields default to the web `DEFAULT_KIOSK_CONFIG` values so a partial persisted
 * payload decodes without error and any missing field is backfilled exactly like the web spread.
 *
 * @property showClock web `config.showClock` — whether the corner clock renders + ticks.
 * @property clockPosition web `config.clockPosition` — the corner wire value, parsed by
 *   [KioskClockPosition.fromWire]; kept as the raw string so the projection runs straight off the JSON.
 * @property dimLevel web `config.dimLevel` — drives the dim-layer opacity (`1 - dimLevel`).
 * @property rotateInterval web `config.rotateInterval` — the rotation cadence in seconds; the dots only
 *   render when it is positive (web `config.rotateInterval > 0`).
 */
@Serializable
data class KioskOverlayConfig(
    val showClock: Boolean = true,
    val clockPosition: String = "bottom-right",
    val dimLevel: Double = DEFAULT_DIM_LEVEL,
    val rotateInterval: Int = DEFAULT_ROTATE_INTERVAL,
) {
    companion object {
        /** Web `DEFAULT_KIOSK_CONFIG.dimLevel`. */
        const val DEFAULT_DIM_LEVEL: Double = 0.5

        /** Web `DEFAULT_KIOSK_CONFIG.rotateInterval` (seconds). */
        const val DEFAULT_ROTATE_INTERVAL: Int = 30
    }
}

/**
 * The fully projected, render-ready view — the native analogue of every conditional the web component
 * evaluates before returning JSX. Pure data (no Compose types) so the projection is unit-tested without a
 * UI host; each field is exactly one web render branch.
 *
 * @property dimAlpha the dim-layer opacity (web `1 - config.dimLevel`, clamped to 0..1) when the layer is
 *   shown, or `null` when `isDimmed` is false and the layer is omitted entirely (web `isDimmed && …`).
 * @property cursorHidden web `isCursorHidden`. The web renders an invisible, aria-hidden CSS-injection layer;
 *   Android has no persistent pointer to hide, so this carries the branch for completeness but drives no
 *   visible surface — faithfully matching the web layer, which is also invisible.
 * @property showClock web `config.showClock` — whether the corner clock renders.
 * @property clockPosition the parsed corner anchor (web `config.clockPosition`).
 * @property showDots web `dashboardCount > 1 && config.rotateInterval > 0` — whether the rotation dots show.
 * @property dotCount the number of dots (web `Array.from({ length: dashboardCount })`).
 * @property activeDotIndex the highlighted dot (web `currentIndex`); when out of range no dot is highlighted,
 *   exactly like the web `i === currentIndex` comparison.
 */
data class KioskOverlayDisplay(
    val dimAlpha: Float?,
    val cursorHidden: Boolean,
    val showClock: Boolean,
    val clockPosition: KioskClockPosition,
    val showDots: Boolean,
    val dotCount: Int,
    val activeDotIndex: Int,
)

/**
 * Pure projection from the surface's props to its render-ready [KioskOverlayDisplay] — a 1:1 port of the
 * conditionals the web component performs: the `1 - config.dimLevel` dim opacity gated by `isDimmed`, the
 * `config.showClock` + `config.clockPosition` clock, and the `dashboardCount > 1 && config.rotateInterval > 0`
 * dots. No formatting or Compose here — only the boolean/numeric derivations.
 */
object KioskOverlayProjection {
    private const val DIM_ALPHA_MIN: Double = 0.0
    private const val DIM_ALPHA_MAX: Double = 1.0

    /** The web threshold `dashboardCount > 1`: a single dashboard shows no rotation indicator. */
    private const val MIN_DASHBOARDS_FOR_DOTS: Int = 1

    /**
     * Select the render-ready view for the given props.
     *
     * @param config the overlay slice of the kiosk config (web `config`).
     * @param isDimmed web `isDimmed` — gates the dim layer.
     * @param isCursorHidden web `isCursorHidden` — carried through (no visible Android surface).
     * @param dashboardCount web `dashboardCount` — the number of rotation dashboards.
     * @param currentIndex web `currentIndex` — the active dashboard index.
     */
    fun project(
        config: KioskOverlayConfig,
        isDimmed: Boolean,
        isCursorHidden: Boolean,
        dashboardCount: Int,
        currentIndex: Int,
    ): KioskOverlayDisplay {
        val dimAlpha =
            if (isDimmed) {
                (DIM_ALPHA_MAX - config.dimLevel).coerceIn(DIM_ALPHA_MIN, DIM_ALPHA_MAX).toFloat()
            } else {
                null
            }
        return KioskOverlayDisplay(
            dimAlpha = dimAlpha,
            cursorHidden = isCursorHidden,
            showClock = config.showClock,
            clockPosition = KioskClockPosition.fromWire(config.clockPosition),
            showDots = dashboardCount > MIN_DASHBOARDS_FOR_DOTS && config.rotateInterval > 0,
            dotCount = dashboardCount.coerceAtLeast(0),
            activeDotIndex = currentIndex,
        )
    }
}

/**
 * Pure, locale- and zone-aware clock formatting — the native analogue of the two `useDateFormat` helpers the
 * web clock calls, `formatTime(now)` and `formatDateWithDay(now)` (web/src/lib/dateFormat.ts). Both run off
 * `java.time` with an injected [Locale] + [ZoneId] so they are deterministic in off-device unit tests; the
 * composable supplies the platform locale (web `useSettings().locale`) and the device zone (the native
 * wall-clock idiom — the web tz preference is a settings concern owned by the page).
 */
object KioskClockFormat {
    // Web `formatDateWithDay` renders `{ weekday: 'short', month: 'short', day: 'numeric' }` — e.g.
    // "Thu, Jun 11". The fixed field order matches the web's explicit Intl options; the names localise.
    private const val DATE_WITH_DAY_PATTERN: String = "EEE, MMM d"

    /**
     * The time text the web renders, `formatTime(now)` — `toLocaleTimeString` with
     * `{ hour: '2-digit', minute: '2-digit' }`. Maps to the platform's localized SHORT time, which honours
     * the locale's 12-/24-hour convention exactly as `toLocaleTimeString` does (e.g. "10:47 PM" in en-US).
     */
    fun time(
        epochMillis: Long,
        locale: Locale,
        zone: ZoneId,
    ): String =
        DateTimeFormatter
            .ofLocalizedTime(FormatStyle.SHORT)
            .withLocale(locale)
            .withZone(zone)
            .format(Instant.ofEpochMilli(epochMillis))

    /**
     * The date text the web renders, `formatDateWithDay(now)` — short weekday + short month + numeric day
     * ("Thu, Jun 11"), locale-aware through [DATE_WITH_DAY_PATTERN].
     */
    fun dateWithDay(
        epochMillis: Long,
        locale: Locale,
        zone: ZoneId,
    ): String =
        DateTimeFormatter
            .ofPattern(DATE_WITH_DAY_PATTERN, locale)
            .withZone(zone)
            .format(Instant.ofEpochMilli(epochMillis))
}

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface [SLUG] — never the
 * clock time, the dashboard index, or the dim level — so a diagnostics line can never leak a user's
 * dashboard posture or local time.
 */
object KioskOverlayDiagnostics {
    /** Diagnostics surface slug emitted with the `view.opened` event. */
    const val SLUG: String = "KioskOverlay"

    private const val VIEW_OPENED = "view.opened"
    private const val SURFACE_KEY = "surface"

    /** Emits the `view.opened` diagnostic for this surface. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to SLUG))
    }
}
